apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts

```diff
@@ -0,0 +1,1029 @@
+      }
+    );
+
+    this._terminal.writeVerboseLine(
+      `${shouldHardlink ? 'Hardlink' : 'Copy from cache'}: ${hardlinkDuration}ms (${hardlinkCount} files)`
+    );
+    //#endregion
+
+    let compilationFailed: boolean = false;
+    if (diagnostics.length > 0) {
+      this._terminal.writeErrorLine(
+        `Encountered ${diagnostics.length} TypeScript error${diagnostics.length > 1 ? 's' : ''}:`
+      );
+      for (const diagnostic of diagnostics) {
+        this._printDiagnosticMessage(ts, diagnostic, true);
+      }
+
+      compilationFailed = true;
+    }
+
+    if (tslintResult?.failures?.length) {
+      this._terminal.writeWarningLine(
+        `Encountered ${tslintResult!.failures.length} TSLint error${
+          tslintResult!.failures.length > 1 ? 's' : ''
+        }:`
+      );
+      for (const tslintFailure of tslintResult!.failures) {
+        const buildFolderRelativeFilename: string = path.relative(
+          this._configuration.buildFolder,
+          tslintFailure.getFileName()
+        );
+        const { line, character } = tslintFailure.getStartPosition().getLineAndCharacter();
+        const severity: string = tslintFailure.getRuleSeverity().toUpperCase();
+        this._terminal.writeWarningLine(
+          '  ',
+          Colors.yellow(`${severity}: ${buildFolderRelativeFilename}(${line + 1},${character + 1})`),
+          Colors.white(' - '),
+          Colors.yellow(`(${tslintFailure.getRuleName()}) ${tslintFailure.getFailure()}`)
+        );
+      }
+    }
+
+    if (compilationFailed) {
+      throw new Error('TypeScript compilation failed.');
+    }
+  }
+
+  private _printDiagnosticMessage(
+    ts: ExtendedTypescript,
+    diagnostic: Typescript.Diagnostic,
+    withIndent: boolean = false
+  ): void {
+    let terminalMessage: (string | IColorableSequence)[];
+    // Code taken from reference example
+    if (diagnostic.file) {
+      const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);
+      const message: string = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
+      const buildFolderRelativeFilename: string = path.relative(
+        this._configuration.buildFolder,
+        diagnostic.file.fileName
+      );
+      terminalMessage = [
+        Colors.red(`ERROR: ${buildFolderRelativeFilename}(${line + 1},${character + 1})`),
+        Colors.white(' - '),
+        Colors.gray(`(TS${diagnostic.code})`),
+        ' ',
+        Colors.red(message)
+      ];
+    } else {
+      terminalMessage = [ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n')];
+    }
+
+    if (withIndent) {
+      terminalMessage = ['  ', ...terminalMessage];
+    }
+
+    switch (diagnostic.category) {
+      case ts.DiagnosticCategory.Error: {
+        this._terminal.writeErrorLine(...terminalMessage);
+        break;
+      }
+
+      case ts.DiagnosticCategory.Warning: {
+        this._terminal.writeWarningLine(...terminalMessage);
+        break;
```

Comment:
Also, continuing the above example log output, the most important information should appear towards the end of the log (i.e. the part that doesn't scroll out of view).  So maybe these details should be moved much earlier (and maybe squashed onto a single line?):

```
Heft version: 0.1.0
Node version: v12.17.0
```

Reply:
There isn't really a more natural place to log this. I think we should spend some time designing the whole logging system, and moving this around can be part of that.